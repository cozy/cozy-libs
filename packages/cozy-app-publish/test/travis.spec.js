/* eslint-env jest */
const path = require('path')

const publishLib = require('../lib/publish')
const postpublish = require('../lib/postpublish')
const prepublish = require('../lib/prepublish')

const mockAppDir = path.join(__dirname, 'mockApps/mockApp')
const mockAppNoEditorDir = path.join(__dirname, 'mockApps/mockAppNoEditor')
const getTravisVariables = require('../utils/getTravisVariables')

jest.mock('../lib/publish', () => jest.fn())
jest.mock('../lib/prepublish', () =>
  jest.fn(options => {
    return Object.assign({}, options, { sha256Sum: 'fakeshasum5644545' })
  })
)
jest.mock('../lib/postpublish', () => jest.fn())

const mockCommons = {
  token: 'registryTokenForTest123',
  slug: 'mock-app',
  commitHash: 'f4a98378271c17e91faa9e70a2718c34c04cfc27',
  buildDir: mockAppDir
}

jest.mock('../utils/getTravisVariables')

const travisScript = require('../lib/travis')

function getOptions(buildUrl = null) {
  const options = {
    spaceName: 'mock_space',
    travis: true,
    buildUrl
  }
  return options
}

describe('Travis publishing script', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    global.console.log = jest.fn()
    // simulate TRAVIS CI environment variables
    getTravisVariables.mockImplementation(() => ({
      TRAVIS_BUILD_DIR: mockCommons.buildDir,
      TRAVIS_TAG: '2.1.8',
      TRAVIS_COMMIT: mockCommons.commitHash,
      TRAVIS_REPO_SLUG: mockCommons.slug,
      // encrypted variables
      REGISTRY_TOKEN: mockCommons.token
    }))
  })

  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('should work correctly if Travis environment variable provided (no TRAVIS_TAG)', async () => {
    getTravisVariables.mockImplementation(() => ({
      TRAVIS_BUILD_DIR: mockCommons.buildDir,
      TRAVIS_TAG: null,
      TRAVIS_COMMIT: mockCommons.commitHash,
      TRAVIS_REPO_SLUG: mockCommons.slug,
      // encrypted variables
      REGISTRY_TOKEN: mockCommons.token
    }))
    jest.spyOn(Date, 'now').mockReturnValue(1551298916519)
    await travisScript(getOptions())
    expect(publishLib).toHaveBeenCalledTimes(1)
    expect(publishLib.mock.calls[0][0]).toMatchSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(1)
    expect(postpublish).toHaveBeenCalledTimes(1)
  })

  it('should work correctly with TRAVIS_TAG', async () => {
    await travisScript(getOptions())
    expect(publishLib).toHaveBeenCalledTimes(1)
    expect(publishLib.mock.calls[0][0]).toMatchSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(1)
    expect(postpublish).toHaveBeenCalledTimes(1)
  })

  it('should work correctly if --build-url provided', async () => {
    await travisScript(getOptions('https://mock/archive/1.0.0.tar.gz'))
    expect(publishLib).toHaveBeenCalledTimes(1)
    expect(publishLib.mock.calls[0][0]).toMatchSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(1)
    expect(postpublish).toHaveBeenCalledTimes(1)
  })

  it('should work correctly if no space name provided', async () => {
    const options = getOptions()
    delete options.spaceName
    await travisScript(options)
    expect(publishLib).toHaveBeenCalledTimes(1)
    expect(publishLib.mock.calls[0][0]).toMatchSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(1)
    expect(postpublish).toHaveBeenCalledTimes(1)
  })

  it('should throw an error if the token is missing', async () => {
    getTravisVariables.mockImplementationOnce(() => ({
      TRAVIS_BUILD_DIR: mockCommons.buildDir,
      TRAVIS_TAG: null,
      TRAVIS_COMMIT: mockCommons.commitHash,
      TRAVIS_REPO_SLUG: mockCommons.slug,
      // encrypted variables
      REGISTRY_TOKEN: ''
    }))
    await expect(
      travisScript(getOptions())
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(0)
    expect(postpublish).toHaveBeenCalledTimes(0)
  })

  it('should throw an error if the editor is missing', async () => {
    getTravisVariables.mockImplementation(() => ({
      TRAVIS_BUILD_DIR: mockAppNoEditorDir,
      TRAVIS_TAG: null,
      TRAVIS_COMMIT: mockCommons.commitHash,
      TRAVIS_REPO_SLUG: mockCommons.slug,
      // encrypted variables
      REGISTRY_TOKEN: mockCommons.token
    }))
    await expect(
      travisScript(getOptions())
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(0)
    expect(postpublish).toHaveBeenCalledTimes(0)
  })

  it('should fail on prePublish error', async () => {
    const options = getOptions()
    prepublish.mockRejectedValueOnce(new Error('Prepublish test error'))
    await expect(travisScript(options)).rejects.toMatchSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(1)
    expect(publishLib).toHaveBeenCalledTimes(0)
    expect(postpublish).toHaveBeenCalledTimes(0)
  })

  it('should fail on publish error', async () => {
    const options = getOptions()
    publishLib.mockRejectedValueOnce(new Error('Publish test error'))
    await expect(travisScript(options)).rejects.toMatchSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(1)
    expect(publishLib).toHaveBeenCalledTimes(1)
    expect(postpublish).toHaveBeenCalledTimes(0)
  })

  it('should fail on postPublish error', async () => {
    const options = getOptions()
    postpublish.mockRejectedValueOnce(new Error('Postpublish test error'))
    await expect(travisScript(options)).rejects.toMatchSnapshot()
    expect(prepublish).toHaveBeenCalledTimes(1)
    expect(publishLib).toHaveBeenCalledTimes(1)
    expect(postpublish).toHaveBeenCalledTimes(1)
  })

  it('should support prefix', async () => {
    const badTag = 'cozy-banks/2.1.8-beta.1'
    const goodTag = 'cozy-drive/2.1.8-beta.2'
    jest.spyOn(global.Date, 'now').mockReturnValue(123456)
    for (const tag of [badTag, goodTag]) {
      publishLib.mockReset()
      getTravisVariables.mockImplementation(() => ({
        TRAVIS_BUILD_DIR: mockCommons.buildDir,
        TRAVIS_TAG: tag,
        TRAVIS_COMMIT: mockCommons.commitHash,
        TRAVIS_REPO_SLUG: mockCommons.slug,
        // encrypted variables
        REGISTRY_TOKEN: mockCommons.token
      }))
      const options = getOptions()
      await travisScript({ ...options, tagPrefix: 'cozy-drive' })
      expect(publishLib).toHaveBeenCalledTimes(1)
      expect(publishLib.mock.calls[0][0]).toMatchSnapshot()
    }
  })
})
