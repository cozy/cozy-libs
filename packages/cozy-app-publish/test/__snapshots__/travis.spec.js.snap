// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Travis publishing script should fail on postPublish error 1`] = `[Error: Postpublish hooks failed: Postpublish test error]`;

exports[`Travis publishing script should fail on prePublish error 1`] = `[Error: Prepublish failed: Prepublish test error]`;

exports[`Travis publishing script should fail on publish error 1`] = `[Error: Publish failed: Publish test error]`;

exports[`Travis publishing script should support prefix 1`] = `
Object {
  "appBuildUrl": "https://github.com/mock-app/archive/f4a98378271c17e91faa9e70a2718c34c04cfc27.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "2.1.8-dev.f4a9837123456",
  "buildCommit": undefined,
  "buildDir": "build",
  "prepublishHook": undefined,
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "fakeshasum5644545",
  "spaceName": "mock_space",
}
`;

exports[`Travis publishing script should support prefix 2`] = `
Object {
  "appBuildUrl": "https://github.com/mock-app/archive/cozy-drive/2.1.8-beta.2.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "cozy-drive/2.1.8-beta.2",
  "buildCommit": undefined,
  "buildDir": "build",
  "prepublishHook": undefined,
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "fakeshasum5644545",
  "spaceName": "mock_space",
}
`;

exports[`Travis publishing script should throw an error if the editor is missing 1`] = `"Registry editor is missing in the manifest. Publishing failed."`;

exports[`Travis publishing script should throw an error if the token is missing 1`] = `"Registry token is missing. Publishing failed."`;

exports[`Travis publishing script should work correctly if --build-url provided 1`] = `
Object {
  "appBuildUrl": "https://mock/archive/1.0.0.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "2.1.8",
  "buildCommit": undefined,
  "buildDir": "build",
  "prepublishHook": undefined,
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "fakeshasum5644545",
  "spaceName": "mock_space",
}
`;

exports[`Travis publishing script should work correctly if Travis environment variable provided (no TRAVIS_TAG) 1`] = `
Object {
  "appBuildUrl": "https://github.com/mock-app/archive/f4a98378271c17e91faa9e70a2718c34c04cfc27.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "2.1.8-dev.f4a98371551298916519",
  "buildCommit": undefined,
  "buildDir": "build",
  "prepublishHook": undefined,
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "fakeshasum5644545",
  "spaceName": "mock_space",
}
`;

exports[`Travis publishing script should work correctly if no space name provided 1`] = `
Object {
  "appBuildUrl": "https://github.com/mock-app/archive/2.1.8.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "2.1.8",
  "buildCommit": undefined,
  "buildDir": "build",
  "prepublishHook": undefined,
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "fakeshasum5644545",
  "spaceName": "default",
}
`;

exports[`Travis publishing script should work correctly with TRAVIS_TAG 1`] = `
Object {
  "appBuildUrl": "https://github.com/mock-app/archive/2.1.8.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "2.1.8",
  "buildCommit": undefined,
  "buildDir": "build",
  "prepublishHook": undefined,
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "fakeshasum5644545",
  "spaceName": "mock_space",
}
`;
